### PR TITLE
Ec extraction: cleanup leak aux variables

### PR DESCRIPTION
**Pull request by cassiersg**
Follow-up to #971 

Improve the handling of leakage aux variables, and also removes the `pop_aux/reset_aux` functions (these seem likely to lead to bugs if forgotten), and replace them with a new function that creates a new env when entering an operation that may use new aux vars (this assumes aux var usage to be properly nested, which currently holds, and seems reasonable).

### CODE REVIEW

File: compiler/src/toEC.ml [hunk 1]
```diff
@@ -468,39 +470,22 @@ let get_funtype env f = snd (Mf.find f env.funs)
...
-  if 0 < i then
-    env.count := Mty.add (prefix,ty) (i-1) !(env.count)
-
-let reset_aux env prefix tys =
-  let do1 ty =
-    env.count:= Mty.add (prefix, ty) 0 !(env.count)
-  in
-  List.iter do1 tys
+  let l = try Mty.find (prefix, ty) !(env.auxv) with Not_found -> [] in
+  env.count := Mty.add (prefix,ty) (i+1) !(env.count);
```
Comment (lyonel2017) [inline compiler/src/toEC.ml:476]: I think increasing the counter of the aux variables should be done only when `fresh` is `true`.

**Review by lyonel2017 — CHANGES_REQUESTED** [review]

File: compiler/src/toEC.ml (hunk 1 above)
Comment (cassiersg) [reply compiler/src/toEC.ml]: indeed

**Comment by cassiersg** [comment]
@bgregoir All `toEC` fixes we discussed are implemented.

File: compiler/src/toEC.ml [hunk 2]
```diff
@@ -357,156 +357,206 @@ type ec_prog = ec_item list
...
 
-let pop_aux env prefix ty =
-  let i = try Mty.find (prefix, ty) !(env.count) with Not_found -> 0 in
-  if 0 < i then
-    env.count := Mty.add (prefix,ty) (i-1) !(env.count)
+  (* Return the last created aux for (prefix, ty) in this env. *)
+  let reuse_aux env prefix ty =
+    let i = Mpty.find (prefix, ty) env.count in
+    let l = Mpty.find (prefix, ty) !(env.auxv) in
+    BatVect.get l (i-1)
```
Comment (lyonel2017) [inline compiler/src/toEC.ml:544]: It might be interesting to add a test for the size of the list `l`, and if the list is empty a call to `create_aux`.

**Review by lyonel2017 — CHANGES_REQUESTED** [review]

File: compiler/src/toEC.ml (hunk 2 above)
Comment (cassiersg) [reply compiler/src/toEC.ml:544]: It('s intentionally not done: if `reuse_aux` is called is called and `create_aux` has not been called before, then it's a bug. It's then better to crash than to (probably) generate incorrect output.

**Review by bgregoir — APPROVED** [review]

File: compiler/src/toEC.ml (hunk 2 above)
Comment (bgregoir) [reply compiler/src/toEC.ml:544]: I am agree with Gaetan

Comment (bgregoir) [reply compiler/src/toEC.ml:544]: Lyonel are you agree to remove this change request ?
